### PR TITLE
decoder: Add support for `LiteralType` as `Constraint`

### PR DESCRIPTION
**Pull request by radeksimko**
Part of https://github.com/hashicorp/terraform-ls/issues/496

--- 

Existing tests were updated as part of https://github.com/hashicorp/hcl-lang/commit/699069384a9c1ef08b7b7a7ab9e8cd6eb8ac739b (not part of this PR) and yield either comparable or better results, except for reference collection, which:

 - we agreed to leave for a separate PR
 - involves other constraints which aren't implemented yet

---

### CODE REVIEW

File: decoder/expr_literal_type_hover.go [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+)
+
+func (lt LiteralType) HoverAtPos(ctx context.Context, pos hcl.Pos) *lang.HoverData {
+	typ := lt.cons.Type
+
+	if typ == cty.DynamicPseudoType {
+		val, _ := lt.expr.Value(nil)
+		if val.IsKnown() && !val.IsNull() {
+			typ = val.Type()
+		}
```
Comment (radeksimko) [inline decoder/expr_literal_type_hover.go:21]:
```suggestion
		val, diags := lt.expr.Value(nil)
		if !diags.HasErrors() {
			typ = val.Type()
		}
```
Should be more or less equivalent, but this is probably more correct.

File: decoder/expr_literal_type_semtok.go [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+)
+
+func (lt LiteralType) SemanticTokens(ctx context.Context) []lang.SemanticToken {
+	typ := lt.cons.Type
+
+	if typ == cty.DynamicPseudoType {
+		val, _ := lt.expr.Value(nil)
+		if val.IsKnown() && !val.IsNull() {
+			typ = val.Type()
+		}
```
Comment (radeksimko) [inline decoder/expr_literal_type_semtok.go:19]:
```suggestion
		val, diags := lt.expr.Value(nil)
		if !diags.HasErrors() {
			typ = val.Type()
		}
```

File: decoder/expr_literal_type_semtok.go [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+			return []lang.SemanticToken{
+				{
+					Type:      lang.TokenString,
+					Modifiers: lang.SemanticTokenModifiers{},
+					Range:     expr.Range(),
+				},
+			}
+		}
+
+		// TODO: report multiline/HEREDOC notation
```
Comment (jpogran) [inline decoder/expr_literal_type_semtok.go:39]: Should this be ticketed?

File: decoder/expr_literal_type_hover.go [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+	"github.com/hashicorp/hcl-lang/lang"
+	"github.com/hashicorp/hcl-lang/schema"
+	"github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/hcl/v2/hclsyntax"
+	"github.com/zclconf/go-cty/cty"
+)
+
+func (lt LiteralType) HoverAtPos(ctx context.Context, pos hcl.Pos) *lang.HoverData {
+	typ := lt.cons.Type
```
Comment (jpogran) [inline decoder/expr_literal_type_hover.go:15]: Is there a case where cons can be nil? Just to guard against a panic here

File: decoder/expr_literal_type_ref_targets.go [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+import (
+	"context"
+
+	"github.com/hashicorp/hcl-lang/reference"
+	"github.com/hashicorp/hcl-lang/schema"
+	"github.com/hashicorp/hcl/v2/hclsyntax"
+)
+
+func (lt LiteralType) ReferenceTargets(ctx context.Context, targetCtx *TargetContext) reference.Targets {
+	typ := lt.cons.Type
```
Comment (jpogran) [inline decoder/expr_literal_type_ref_targets.go:12]: Is there a case where cons can be nil? Just to guard against a panic here

File: decoder/expr_literal_type_semtok.go [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+	"context"
+
+	"github.com/hashicorp/hcl-lang/lang"
+	"github.com/hashicorp/hcl-lang/schema"
+	"github.com/hashicorp/hcl/v2/hclsyntax"
+	"github.com/zclconf/go-cty/cty"
+)
+
+func (lt LiteralType) SemanticTokens(ctx context.Context) []lang.SemanticToken {
+	typ := lt.cons.Type
```
Comment (jpogran) [inline decoder/expr_literal_type_semtok.go:13]: Is there a case where cons can be nil? Just to guard against a panic here

**Review by jpogran — APPROVED** [review]
Small concerns. Great work!

File: decoder/expr_literal_type_ref_targets.go (hunk 5 above)
Comment (radeksimko) [reply decoder/expr_literal_type_ref_targets.go:12]: It's a good point, although that would indicate invalid schema. So I think the best place to cover that would be in the schema validation. I will add some checks there!

File: decoder/expr_literal_type_semtok.go (hunk 6 above)
Comment (radeksimko) [reply decoder/expr_literal_type_semtok.go:13]: It's a good point, although that would indicate invalid schema. So I think the best place to cover that would be in the schema validation. I will add some checks there!

**Comment by radeksimko** [comment]
I added `LiteralType` validation for nil `Type` and also another one to ensure there's either `Constraint` or `Expr`, so that should ensure integrity on a higher level and ensure there's no `nil` `Constraint`.

If we still somehow end up with `nil` constraint further down then I would call that a really bad bug which needs to be fixed in the place that produces the schema, not on our side. In other words, we shouldn't be producing invalid schema in the first place anywhere and then hiding that fact, so crash is the right thing to do in such a case.

**Pull request merged**